ocaml/xapi/xapi_pool_update.ml

```diff
@@ -532,21 +532,42 @@ let resync_host ~__context ~host =
     |> List.iter (fun self -> destroy ~__context ~self)
   end
 
-let path_from_uri uri =
+let path_and_host_from_uri uri =
   (* remove any dodgy use of "." or ".." NB we don't prevent the use of symlinks *)
-  String.sub_to_end uri (String.length Constants.get_pool_update_download_uri)
-  |> Filename.concat !Xapi_globs.host_update_dir
-  |> Uri.pct_decode
-  |> Stdext.Unixext.resolve_dot_and_dotdot
+  let host_and_path = String.sub_to_end uri (String.length Constants.get_pool_update_download_uri) in
+  match String.split ~limit:2 '/' host_and_path with
+  | [host; untrusted_path] ->
+    let resolved_path = untrusted_path
+      |> Filename.concat !Xapi_globs.host_update_dir
+      |> Uri.pct_decode
+      |> Stdext.Unixext.resolve_dot_and_dotdot
+    in
+    (host,resolved_path)
+  | _ -> failwith "Expecting "
+
+let proxy_request req s host_uuid =
+  Server_helpers.exec_with_new_task "pool_update_proxy_request" (fun __context ->
+    let host = Db.Host.get_by_uuid ~__context ~uuid:host_uuid in
+    let ip = Db.Host.get_address ~__context ~self:host in
+    let transport = Xmlrpc_client.(SSL(SSL.make (), ip, 443)) in
+    Xmlrpc_client.with_transport transport (fun fd ->
+      Unixext.really_write_string fd (Http.Request.to_wire_string req);
+      Unixext.proxy s fd)
```

Comment:
We have to mark the request closed here by `req.Request.close <- true` so Http server will close the socket, otherwise the sock will be left open until client close the other peer.